pkg/retriever/bitswaphelpers/multiblockstore.go

```diff
@@ -0,0 +1,146 @@
+package bitswaphelpers
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-libipfs/blocks"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+// ErrNotSupported indicates an operation not supported by the MultiBlockstore
+var ErrNotSupported = errors.New("not supported")
+
+// ErrAlreadyRegistered means something has already been registered for a retrieval id
+var ErrAlreadyRegisterd = errors.New("already registered")
+
+// ErrAlreadyRegistered means there is nothing registered for a retrieval id
+var ErrNotRegistered = errors.New("not registered")
+
+// MultiBlockstore creates a blockstore based on one or more linkystems, extracting the target linksystem for each request
+// from the retrieval id context key
+type MultiBlockstore struct {
+	linkSystems   map[types.RetrievalID]*linking.LinkSystem
+	linkSystemsLk sync.RWMutex
+}
+
+// NewMultiblockstore returns a new MultiBlockstore
+func NewMultiblockstore() *MultiBlockstore {
+	return &MultiBlockstore{
+		linkSystems: make(map[types.RetrievalID]*linking.LinkSystem),
+	}
+}
+
+// AddLinkSystem registers a linksystem to use for a given retrieval id
+func (mbs *MultiBlockstore) AddLinkSystem(id types.RetrievalID, lsys *linking.LinkSystem) error {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	if _, ok := mbs.linkSystems[id]; ok {
+		return ErrAlreadyRegisterd
+	}
+	mbs.linkSystems[id] = lsys
+	return nil
+}
+
+// RemoveLinkSystem unregisters the link system for a given retrieval id
+func (mbs *MultiBlockstore) RemoveLinkSystem(id types.RetrievalID) {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	delete(mbs.linkSystems, id)
+}
+
+// DeleteBlock is not supported
+func (mbs *MultiBlockstore) DeleteBlock(ctx context.Context, c cid.Cid) error {
+	return ErrNotSupported
+}
+
+// Has is not supported
+func (mbs *MultiBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	return false, ErrNotSupported
+}
+
+type byteReader interface {
+	Bytes() []byte
+}
+
+// Get returns a block only if the given ctx contains a retrieval ID as a value that
+// references a known linksystem. If it does, it uses that linksystem to load the block
+func (mbs *MultiBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	id, err := types.RetrievalIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	mbs.linkSystemsLk.RLock()
+	lsys, ok := mbs.linkSystems[id]
+	mbs.linkSystemsLk.RUnlock()
+	if !ok {
+		return nil, ErrNotRegistered
+	}
+	r, err := lsys.StorageReadOpener(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: c})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I suspect you're not going to like me for this .... but if we want to convert back to a classic Blockstore then we need to "correct" this one too because I eschewed pulling in go-block-format in the go-car changes. See note about it under "Secondary items" in https://github.com/ipld/go-car/pull/363. Also the note I just dropped in https://github.com/ipld/go-ipld-prime/pull/493#issuecomment-1430721773

```suggestion
	if err != nil {
		if nf, ok := err.(interface{ NotFound() bool }); ok && nf.NotFound() {
			return nil, format.ErrNotFound{Cid: c}
		}
		return nil, err
	}
```

Reply:
it's fine -- I was absolutely worried about this anyway.